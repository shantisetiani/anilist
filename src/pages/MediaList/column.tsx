import { Media, MediaList } from "generated/graphql";
import { Link } from "react-router-dom";
import { ROUTES } from "routes";

export const columns = [
  {
    title: "No",
    key: "no",
    render: (data: MediaList, row: any, index: number) => index + 1,
  },
  {
    title: "Title",
    dataIndex: "media",
    key: "title",
    render: (media: Media, row: any) => (
      <Link to={ROUTES.MEDIA_DETAIL} state={{ id: row.id }}>
        {media.title?.userPreferred}
      </Link>
    ),
  },
  {
    title: "Type",
    dataIndex: "media",
    key: "type",
    render: (media: Media) => media.type,
  },
  {
    title: "Genre",
    dataIndex: "media",
    key: "genre",
    render: (media: Media) => media.genres?.join(", "),
  },
  {
    title: "Status",
    dataIndex: "status",
    key: "status",
  },
  {
    title: "Episode",
    key: "episode",
    render: (row: MediaList) =>
      `${row.progress}/${row?.media?.episodes ? row.media.episodes : "?"}`,
  },
];
