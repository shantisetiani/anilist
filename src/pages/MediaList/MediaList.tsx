import React, { Dispatch, SetStateAction } from "react";
import {
  MediaListQuery,
  MediaListSort,
  MediaType,
  MediaListStatus,
} from "generated/graphql";
import { Row, Col, Table, Spin, Card, Select } from "antd";
import { capitalizeFirstLetter } from "utilities/string";
import { columns } from "./column";

interface MediaListProps {
  mediaList: MediaListQuery | undefined;
  loading: boolean;
  mediaFilter: MediaFilter;
  setMediaFilter: Dispatch<SetStateAction<MediaFilter>>;
}

export interface MediaFilter {
  sort: MediaListSort | null;
  type: MediaType;
  status: MediaListStatus;
}

const MediaList: React.FC<MediaListProps> = ({
  mediaList,
  loading,
  mediaFilter,
  setMediaFilter,
}) => {
  const statusList = Object.values(MediaListStatus);

  /* Handle media type filter on change */
  const onMediaTypeChange = (value: MediaType) => {
    setMediaFilter({ ...mediaFilter, sort: null, type: value });
  };

  /* Handle media status filter on change */
  const onMediaStatusChange = (value: MediaListStatus) => {
    setMediaFilter({ ...mediaFilter, sort: null, status: value });
  };

  const renderFilter = () => (
    <Card>
      <Row gutter={16}>
        <Col span={12}>
          <div>Media Type</div>
          <Select
            style={{ width: "100%" }}
            placeholder="Select Media Type"
            onChange={onMediaTypeChange}
          >
            <Select.Option value={MediaType.Anime}>Anime</Select.Option>
            <Select.Option value={MediaType.Manga}>Manga</Select.Option>
          </Select>
        </Col>
        <Col span={12}>
          <div>Media Status</div>
          <Select
            showSearch
            style={{ width: "100%" }}
            placeholder="Select Media Status"
            onChange={onMediaStatusChange}
          >
            {statusList.map((status) => (
              <Select.Option value={status} key={status}>
                {capitalizeFirstLetter(status)}
              </Select.Option>
            ))}
          </Select>
        </Col>
      </Row>
    </Card>
  );

  return (
    <div>
      <Row>
        <Col span={24}>
          <h1>Media List</h1>
        </Col>
      </Row>
      <Row>
        <Col span={24}>{renderFilter()}</Col>
      </Row>
      <Row style={{ marginTop: "24px" }}>
        <Col span={24}>
          {loading ? (
            <div className="text-center">
              <Spin size="large" />
            </div>
          ) : (
            <Table
              columns={columns}
              dataSource={mediaList?.MediaList ? [mediaList.MediaList] : []}
              bordered={true}
              scroll={{ x: "auto" }}
            />
          )}
        </Col>
      </Row>
    </div>
  );
};

export default MediaList;
