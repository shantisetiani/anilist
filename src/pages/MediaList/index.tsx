import React, { useState, useContext, useEffect } from "react";
import { useMediaListQuery, MediaListSort } from "generated/graphql";
import { AlertContext } from "context/alert";
import MediaList, { MediaFilter } from "./MediaList";

const MediaListContainer: React.FC = () => {
  /* Filter's state */
  const [mediaFilter, setMediaFilter] = useState<MediaFilter>({
    sort: MediaListSort.MediaPopularityDesc,
  } as MediaFilter);

  /* Get data */
  const {
    data: mediaList,
    error: listError,
    loading: listLoading,
  } = useMediaListQuery({
    variables: {
      sort: mediaFilter.sort,
      type: mediaFilter.type,
      status: mediaFilter.status,
    },
  });

  const alertContext = useContext(AlertContext);

  /* Show error alert */
  useEffect(() => {
    if (listError) {
      alertContext.setAlert(true);
    }
    // eslint-disable-next-line
  }, [listError]);

  return (
    <MediaList
      mediaList={mediaList}
      loading={listLoading}
      mediaFilter={mediaFilter}
      setMediaFilter={setMediaFilter}
    />
  );
};

export default MediaListContainer;
