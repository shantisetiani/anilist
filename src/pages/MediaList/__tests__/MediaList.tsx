import TestRenderer from "react-test-renderer";
import { MockedProvider } from "@apollo/react-testing";
import { useMediaListQuery } from "generated/graphql";
import MediaList from "pages/MediaList/MediaList";

const mocks: any = [
  {
    request: {
      query: useMediaListQuery,
      variables: {
        type: "ANIME",
        status: "PLANNING",
      },
    },
    result: {
      data: {
        mediaList: { id: "1", title: "Shingeki no Kyojin" },
      },
    },
  },
];

describe("MediaList", () => {
  it("Should render loading state", async () => {
    const mockMediaFilter = {} as any;
    const mockSetMediaFilter = {} as any;

    const component = TestRenderer.create(
      <MockedProvider mocks={[]}>
        <MediaList
          mediaList={{}}
          loading={false}
          mediaFilter={mockMediaFilter}
          setMediaFilter={mockSetMediaFilter}
        />
      </MockedProvider>
    ).toJSON();
    expect(component).toMatchSnapshot();
  });
});
