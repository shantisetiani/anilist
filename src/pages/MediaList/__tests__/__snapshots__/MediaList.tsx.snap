// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaList Should render loading state 1`] = `
<div>
  <div
    className="ant-row"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-24"
      style={Object {}}
    >
      <h1>
        Media List
      </h1>
    </div>
  </div>
  <div
    className="ant-row"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-24"
      style={Object {}}
    >
      <div
        className="ant-card ant-card-bordered"
      >
        <div
          className="ant-card-body"
          style={Object {}}
        >
          <div
            className="ant-row"
            style={
              Object {
                "marginLeft": -8,
                "marginRight": -8,
              }
            }
          >
            <div
              className="ant-col ant-col-12"
              style={
                Object {
                  "paddingLeft": 8,
                  "paddingRight": 8,
                }
              }
            >
              <div>
                Media Type
              </div>
              <div
                className="ant-select ant-select-single ant-select-show-arrow"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                <div
                  className="ant-select-selector"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                >
                  <span
                    className="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="undefined_list_0"
                      aria-autocomplete="list"
                      aria-controls="undefined_list"
                      aria-haspopup="listbox"
                      aria-owns="undefined_list"
                      autoComplete="off"
                      className="ant-select-selection-search-input"
                      onChange={[Function]}
                      onCompositionEnd={[Function]}
                      onCompositionStart={[Function]}
                      onKeyDown={[Function]}
                      onMouseDown={[Function]}
                      onPaste={[Function]}
                      readOnly={true}
                      role="combobox"
                      style={
                        Object {
                          "opacity": 0,
                        }
                      }
                      type="search"
                      unselectable="on"
                      value=""
                    />
                  </span>
                  <span
                    className="ant-select-selection-placeholder"
                  >
                    Select Media Type
                  </span>
                </div>
                <span
                  aria-hidden={true}
                  className="ant-select-arrow"
                  onMouseDown={[Function]}
                  style={
                    Object {
                      "WebkitUserSelect": "none",
                      "userSelect": "none",
                    }
                  }
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    className="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
            <div
              className="ant-col ant-col-12"
              style={
                Object {
                  "paddingLeft": 8,
                  "paddingRight": 8,
                }
              }
            >
              <div>
                Media Status
              </div>
              <div
                className="ant-select ant-select-single ant-select-show-arrow ant-select-show-search"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                <div
                  className="ant-select-selector"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                >
                  <span
                    className="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="undefined_list_0"
                      aria-autocomplete="list"
                      aria-controls="undefined_list"
                      aria-haspopup="listbox"
                      aria-owns="undefined_list"
                      autoComplete="off"
                      className="ant-select-selection-search-input"
                      onChange={[Function]}
                      onCompositionEnd={[Function]}
                      onCompositionStart={[Function]}
                      onKeyDown={[Function]}
                      onMouseDown={[Function]}
                      onPaste={[Function]}
                      readOnly={false}
                      role="combobox"
                      style={
                        Object {
                          "opacity": null,
                        }
                      }
                      type="search"
                      unselectable={null}
                      value=""
                    />
                  </span>
                  <span
                    className="ant-select-selection-placeholder"
                  >
                    Select Media Status
                  </span>
                </div>
                <span
                  aria-hidden={true}
                  className="ant-select-arrow"
                  onMouseDown={[Function]}
                  style={
                    Object {
                      "WebkitUserSelect": "none",
                      "userSelect": "none",
                    }
                  }
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    className="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row"
    style={
      Object {
        "marginTop": "24px",
      }
    }
  >
    <div
      className="ant-col ant-col-24"
      style={Object {}}
    >
      <div
        className="ant-table-wrapper"
      >
        <div
          className="ant-spin-nested-loading"
        >
          <div
            className="ant-spin-container"
          >
            <div
              className="ant-table ant-table-bordered ant-table-empty ant-table-scroll-horizontal"
            >
              <div
                className="ant-table-container"
              >
                <div
                  className="ant-table-content"
                  onScroll={[Function]}
                  style={
                    Object {
                      "overflowX": "auto",
                      "overflowY": "hidden",
                    }
                  }
                >
                  <table
                    style={
                      Object {
                        "minWidth": "100%",
                        "tableLayout": "auto",
                        "width": "auto",
                      }
                    }
                  >
                    <colgroup />
                    <thead
                      className="ant-table-thead"
                    >
                      <tr>
                        <th
                          className="ant-table-cell"
                          colSpan={null}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          rowSpan={null}
                          style={Object {}}
                        >
                          No
                        </th>
                        <th
                          className="ant-table-cell"
                          colSpan={null}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          rowSpan={null}
                          style={Object {}}
                        >
                          Title
                        </th>
                        <th
                          className="ant-table-cell"
                          colSpan={null}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          rowSpan={null}
                          style={Object {}}
                        >
                          Type
                        </th>
                        <th
                          className="ant-table-cell"
                          colSpan={null}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          rowSpan={null}
                          style={Object {}}
                        >
                          Genre
                        </th>
                        <th
                          className="ant-table-cell"
                          colSpan={null}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          rowSpan={null}
                          style={Object {}}
                        >
                          Status
                        </th>
                        <th
                          className="ant-table-cell"
                          colSpan={null}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          rowSpan={null}
                          style={Object {}}
                        >
                          Episode
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      className="ant-table-tbody"
                    >
                      <tr
                        aria-hidden="true"
                        className="ant-table-measure-row"
                        style={
                          Object {
                            "fontSize": 0,
                            "height": 0,
                          }
                        }
                      >
                        <td
                          style={
                            Object {
                              "border": 0,
                              "height": 0,
                              "padding": 0,
                            }
                          }
                        >
                          <div
                            style={
                              Object {
                                "height": 0,
                                "overflow": "hidden",
                              }
                            }
                          >
                             
                          </div>
                        </td>
                        <td
                          style={
                            Object {
                              "border": 0,
                              "height": 0,
                              "padding": 0,
                            }
                          }
                        >
                          <div
                            style={
                              Object {
                                "height": 0,
                                "overflow": "hidden",
                              }
                            }
                          >
                             
                          </div>
                        </td>
                        <td
                          style={
                            Object {
                              "border": 0,
                              "height": 0,
                              "padding": 0,
                            }
                          }
                        >
                          <div
                            style={
                              Object {
                                "height": 0,
                                "overflow": "hidden",
                              }
                            }
                          >
                             
                          </div>
                        </td>
                        <td
                          style={
                            Object {
                              "border": 0,
                              "height": 0,
                              "padding": 0,
                            }
                          }
                        >
                          <div
                            style={
                              Object {
                                "height": 0,
                                "overflow": "hidden",
                              }
                            }
                          >
                             
                          </div>
                        </td>
                        <td
                          style={
                            Object {
                              "border": 0,
                              "height": 0,
                              "padding": 0,
                            }
                          }
                        >
                          <div
                            style={
                              Object {
                                "height": 0,
                                "overflow": "hidden",
                              }
                            }
                          >
                             
                          </div>
                        </td>
                        <td
                          style={
                            Object {
                              "border": 0,
                              "height": 0,
                              "padding": 0,
                            }
                          }
                        >
                          <div
                            style={
                              Object {
                                "height": 0,
                                "overflow": "hidden",
                              }
                            }
                          >
                             
                          </div>
                        </td>
                      </tr>
                      <tr
                        className="ant-table-placeholder"
                        style={
                          Object {
                            "display": null,
                          }
                        }
                      >
                        <td
                          className="ant-table-cell"
                          colSpan={6}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          rowSpan={null}
                          style={Object {}}
                        >
                          <div
                            className="ant-table-expanded-row-fixed"
                            style={
                              Object {
                                "left": 0,
                                "overflow": "hidden",
                                "position": "sticky",
                                "width": 0,
                              }
                            }
                          >
                            <div
                              className="ant-empty ant-empty-normal"
                            >
                              <div
                                className="ant-empty-image"
                              >
                                <svg
                                  className="ant-empty-img-simple"
                                  height="41"
                                  viewBox="0 0 64 41"
                                  width="64"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <g
                                    fill="none"
                                    fillRule="evenodd"
                                    transform="translate(0 1)"
                                  >
                                    <ellipse
                                      className="ant-empty-img-simple-ellipse"
                                      cx="32"
                                      cy="33"
                                      rx="32"
                                      ry="7"
                                    />
                                    <g
                                      className="ant-empty-img-simple-g"
                                      fillRule="nonzero"
                                    >
                                      <path
                                        d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                      />
                                      <path
                                        className="ant-empty-img-simple-path"
                                        d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </div>
                              <div
                                className="ant-empty-description"
                              >
                                No Data
                              </div>
                            </div>
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
